File: tests/party_reaction.py
"""Readonly Q native evidence. Prior drawn facing is not an inferred attacker direction.
Actual cell fingerprints refer to draw history, never to a synchronized screenshot.
"""
import json,re
from pathlib import Path
from party_combat import integer,finite,vec,clip_frame
GOLDEN=json.loads((Path(__file__).parent/'fixtures/party-reaction-frames.json').read_text())
OBSERVE_SCRIPT="""()=>{const a=window.__CHRONO_TEST__;return {state:a.snapshot(),motion:a.view().partyCombat.reactions};}"""
DURATIONS=[90,90,100,130]
def hurt_frame(tick,start):
    ms=(tick/60-start/60)*1000;edge=0
    for i,d in enumerate(DURATIONS):
        edge+=d
        if 0<=ms<edge:return i
    return None

def assert_reaction_history(record,*,require_hit=False):
    s=record['state'];m=record['motion'];tick=s['ticks']
    assert integer(tick) and m['tick']==tick and m['chapter']==s['chapter'] and s['chapter']!='lab'
    assert m['profile']=='vq03q-party-reaction-continuity' and m['disposed'] is False
    assert m['historyLimit']==24 and integer(m['historyDropped']) and len(m['history'])<=24
    assert m['textureReadFailures']==0 and m['stateMutation'] is False and m['additionalGpuResources']==0 and m['approved'] is False
    assert type(m['reducedMotion']) is bool
    groups={};different=0;last=-1
    for h in m['history']:
        slot=h['slot'];t=h['tick'];c=h['cause'];p=c['beforeDraw'];e=c['effect'];d=h['facing'];f=h['frame'];hero=h['actor']
        assert type(slot) is int and 0<=slot<=2 and integer(t) and last<=t<=tick;last=t
        assert hero in (['crono'] if slot==0 else ['marle','lucca'] if slot==1 else ['frog','marle'])
        assert type(d) is int and 0<=d<4 and type(f) is int and 0<=f<4
        assert type(h['fallbackFacing']) is int and 0<=h['fallbackFacing']<4
        assert h['pose']=='hurt' and finite(h['hp']) and h['hp']>0 and h['mode'] in ('battle','victory')
        assert h['scope']=='draw-texture-not-framebuffer' and type(h['reducedMotion']) is bool
        assert c['kind']=='delivered-target-hit' and integer(c['tick']) and integer(p['tick']) and p['tick']<=c['tick']<=t
        assert p['facing']==d and e['kind']=='hit' and 'actor' not in e and e.get('guest') is not True
        assert vec(c['target']) and c['target']=={'x':e['x'],'z':e['z']}
        assert f==hurt_frame(t,c['tick'])
        assert h['texture']=={'width':48,'height':64,'fnv1a32':GOLDEN['cells'][f'{hero}/hurt/{d}/{f}']['fnv1a32']}
        key=(slot,c['tick'])
        if key not in groups:groups[key]={'cause':c,'frames':set()}
        assert groups[key]['cause']==c
        groups[key]['frames'].add(f);different+=h['fallbackFacing']!=d
    if require_hit:assert groups,'No real delivered party reaction observed'
    return {'deliveredHits':len(groups),'reactionFrameSamples':len(m['history']),'completeObservedReactions':sum(g['frames']=={0,1,2,3} for g in groups.values()),'differentFallbackSamples':different}

def assert_reaction_report(report,expected_sha,*,expected_build=('0.9.64','VQ03Q')):
    assert re.fullmatch('[0-9a-f]{40}',expected_sha)
    assert report['sourceSha']==report['build']['sourceSha']==expected_sha
    assert (report['build']['version'],report['build']['batch'])==expected_build
    assert report['status']=='passed' and report['nativeInputsOnly'] is True and report['errors']==[]
    assert report['physicalDevice'] is False and report['fullAnimationComplete'] is False and report['artApproved'] is False and report['framebufferSynchronized'] is False
    return assert_reaction_history(report['observation'],require_hit=True)

def observe_reaction_suffix(page,out,build,source_sha,errors,*,expected_build=('0.9.64','VQ03Q')):
    """Append only reads after every original N/O/P input, wait, capture and assertion."""
    r={'status':'running','sourceSha':source_sha,'build':build,'nativeInputsOnly':True,'physicalDevice':False,'fullAnimationComplete':False,'artApproved':False,'framebufferSynchronized':False}
    try:
        v=page.evaluate(OBSERVE_SCRIPT);r['observation']=v
        (out/'party-reaction-observation.json').write_text(json.dumps(v,ensure_ascii=False,indent=2),encoding='utf-8')
        assert not errors,errors
        r.update(status='passed',errors=list(errors))
        r['positiveSamples']=assert_reaction_report(r,source_sha,expected_build=expected_build)
        r['limitations']=['Only delivered target hits and previously drawn facing; no attacker inferred from target-only data.','History is actual retained canvas cells, not synchronous framebuffer or full animation approval.','No original route, input, wait, capture or threshold changed. Missing native selector-change and down coverage remains open.']
    except Exception as exc:
        r.update(status='failed',failure=str(exc),errors=list(errors));raise
    finally:(out/'party-reaction-report.json').write_text(json.dumps(r,ensure_ascii=False,indent=2),encoding='utf-8')

File: src/party-reaction.ts
import type {State,Effect} from './core';
import {CLIP_MS} from './hero-art';
import type {PoseSample} from './pose-player';

/** Retain the previously drawn direction for a real delivered target hit.
 * This never guesses an attacker, triggers a pose, or changes simulation state. */
export const PARTY_REACTION=Object.freeze({id:'vq03q-party-reaction-continuity',historyLimit:24,approved:false});
type Owner={actor:State['players'][number];key:string};
type Draw={tick:number;facing:number};
type Cause={kind:'delivered-target-hit';tick:number;target:{x:number;z:number};beforeDraw:Draw;effect:Effect};
type Cell={width:number;height:number;fnv1a32:number};
type Row={slot:number;actor:string;tick:number;hp:number;mode:State['mode'];pose:'hurt';frame:number;facing:number;fallbackFacing:number;reducedMotion:boolean;cause:Cause;texture:Cell|null;scope:'draw-texture-not-framebuffer'};
type Result=PoseSample&{facing:number};
const direction=(v:number)=>Number.isInteger(v)&&v>=0&&v<4;
const duration=CLIP_MS.hurt.reduce((n,t)=>n+t,0);
export class PartyReactionMotion {
 private state:State|null=null;
 private chapter='';
 private tick:number|null=null;
 private reduced=false;
 private disposed=false;
 private owners:(Owner|null)[]=[null,null,null];
 private drawn:(Draw|null)[]=[null,null,null];
 private staged:(Draw|null)[]=[null,null,null];
 private locks:(Cause|null)[]=[null,null,null];
 private pending:(Row|null)[]=[null,null,null];
 private keys:(string|null)[]=[null,null,null];
 private history:Row[]=[];
 private dropped=0;
 private readFailures=0;
 begin(s:State,reduced:boolean,owners:(Owner|null)[]):void{
  if(this.disposed)return;
  if(this.state!==s||this.chapter!==s.chapter||(this.tick!==null&&s.ticks<this.tick))this.reset();
  this.state=s;this.chapter=s.chapter;this.tick=s.ticks;this.reduced=reduced;
  for(let i=0;i<3;i++){
   const owner=owners[i]??null,old=this.owners[i];
   if(!owner||old?.actor!==owner.actor||old?.key!==owner.key||owner.actor.hp<=0||s.chapter==='lab'||(s.mode!=='battle'&&s.mode!=='victory'))this.clearSlot(i);
   this.owners[i]=owner;
  }
 }
 private clearSlot(i:number):void{this.drawn[i]=null;this.staged[i]=null;this.locks[i]=null;this.pending[i]=null;this.keys[i]=null;}
 receive(s:State,e:Effect):void{
  if(this.disposed||s!==this.state||s.chapter==='lab'||s.mode!=='battle'||e.kind!=='hit'||e.actor!==undefined||e.guest===true||!Number.isFinite(e.x)||!Number.isFinite(e.z))return;
  // Exact, unique living target only. Near or overlapping actors are not disambiguated.
  const matches=this.owners.flatMap((o,i)=>o&&o.actor.hp>0&&o.actor.x===e.x&&o.actor.z===e.z?[i]:[]);
  if(matches.length!==1)return;
  const i=matches[0]!,before=this.drawn[i];
  if(!before||before.tick>s.ticks||!direction(before.facing)||this.locks[i]?.tick===s.ticks)return;
  this.locks[i]={kind:'delivered-target-hit',tick:s.ticks,target:{x:e.x,z:e.z},beforeDraw:{...before},effect:structuredClone(e)};
 }
 sample(slot:number,result:Result):number{
  const owner=this.owners[slot],s=this.state;let facing=result.facing;
  if(this.disposed||!owner||!s||this.tick===null)return facing;
  const lock=this.locks[slot];
  if(lock){
   const age=(this.tick/60-lock.tick/60)*1000;
   // Pose ownership stays with the existing ActorTimeline. An interrupted reaction
   // cannot resume later simply because another hurt pose happens to be sampled.
   if(owner.actor.hp<=0||result.pose!=='hurt'||age<0||age>=duration)this.locks[slot]=null;
   else {
    facing=lock.beforeDraw.facing;
    const key=`${lock.tick}/${result.frame}/${facing}/${this.reduced}`;
    if(key!==this.keys[slot]){
     this.keys[slot]=key;
     const row:Row={slot,actor:owner.key,tick:this.tick,hp:owner.actor.hp,mode:s.mode,pose:'hurt',frame:result.frame,facing,fallbackFacing:result.facing,reducedMotion:this.reduced,cause:structuredClone(lock),texture:null,scope:'draw-texture-not-framebuffer'};
     this.pending[slot]=row;this.history.push(row);
     if(this.history.length>PARTY_REACTION.historyLimit){this.history.shift();this.dropped++;}
    }
   }
  }
  if(!this.locks[slot]){this.keys[slot]=null;this.pending[slot]=null;}
  this.staged[slot]=owner.actor.hp>0&&s.chapter!=='lab'&&(s.mode==='battle'||s.mode==='victory')&&direction(facing)?{tick:this.tick,facing}:null;
  return facing;
 }
 /** Called by the retained renderer after the real draw. No per-frame readback;
  * fingerprint only a newly observed reaction cell, not the framebuffer. */
 recordDraw(slot:number,read:()=>{width:number;height:number;data:ArrayLike<number>}):void{
  if(this.disposed)return;
  const staged=this.staged[slot];if(staged){this.drawn[slot]={...staged};this.staged[slot]=null;}
  const row=this.pending[slot];if(!row)return;this.pending[slot]=null;
  try{
   const p=read();if(p.width!==48||p.height!==64||p.data.length!==48*64*4)throw Error('Unexpected reaction cell');
   let h=2166136261;for(let i=0;i<p.data.length;i++)h=Math.imul(h^p.data[i]!,16777619);
   row.texture={width:p.width,height:p.height,fnv1a32:h>>>0};
  }catch{this.readFailures++;}
 }
 reset():void{this.state=null;this.chapter='';this.tick=null;this.owners.fill(null);this.drawn.fill(null);this.staged.fill(null);this.locks.fill(null);this.pending.fill(null);this.keys.fill(null);this.history=[];this.dropped=0;this.readFailures=0;}
 dispose():void{this.reset();this.disposed=true;}
 inspect(){return {profile:PARTY_REACTION.id,tick:this.tick,chapter:this.chapter,disposed:this.disposed,reducedMotion:this.reduced,historyLimit:PARTY_REACTION.historyLimit,historyDropped:this.dropped,textureReadFailures:this.readFailures,history:structuredClone(this.history),stateMutation:false,additionalGpuResources:0,approved:false};}
}
